fastascripts â€“ Manipulate FASTA files
=====================================

These scripts were never meant to see the light of day. I apologize for the
distasteful code.

Utilities
=========

* `facanon`
	Output the lexicographically smaller of the sequence and its reverse complement
* `facat`
	Concatenate FASTA files and add a prefix to the identifier
* `fachain`
	Merge overlapping sequences
* `faclean`
	Reformat and optionally remove short contigs from a FASTA file
* `facmp`
	Compare pairs of sequences
* `facstont`
	Convert colour-space sequence to nucleotides
* `fadecimate`
	Randomly keep 1 in every N pairs of reads
* `faget`
	Select sequences from a FASTA file by identifier
* `fagrep`
	Search a FASTA file using a regular expression
* `fakmer`
	Generate tiled k-mers
* `falength`
	Print the lengths of sequences
* `falint`
	Check the syntax of a FASTA file
* `famd5`
	Calculate a MD5 digest for a FASTA file
* `famd5seq`
	Calculate a MD5 digest for each sequence
* `fanttocs`
	Convert nucleotides to colour-space sequence
* `farand`
	Generate a FASTA file with random sequence
* `farc`
	Reverse and complement the sequences
* `farenumber`
	Renumber the sequences
* `faseperate-mates`
	Separate paired reads into two files
* `fasplit-read`
	Split a read into two at the midpoint
* `fastqtofa`
	Convert a FASTQ file to FASTA format
* `fatoagp`
	Convert FASTA scaffolds to FASTA contigs and an AGP file
* `fatofastq`
	 Convert a FASTA file to a FASTQ file
* `fatoseq`
	Remove FASTA headers
* `faunamb`
	Convert IUPAC-IUB ambiguity codes to ACGT
* `faunscaffold`
	Break scaffolds into contigs at Ns
